web/src/lottie-player.ts

```diff
@@ -0,0 +1,577 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+import { html, PropertyValueMap, LitElement, type TemplateResult } from 'lit';
+import { customElement, property } from 'lit/decorators.js';
+
+// @ts-ignore: WASM Glue code doesn't have type & Only available on build progress
+import Module from '../dist/thorvg-wasm';
+import { THORVG_VERSION } from './version';
+
+let _tvg: any;
+let _module: any;
+(async () => {  
+  _module = await Module();
+  _tvg = new _module.TvgLottieAnimation();
+})();
+
+// Define library version
+export interface LibraryVersion {
+  THORVG_VERSION: string
+}
+
+// Define file type which can be exported
+export enum ExportableType {
+  GIF = 'gif',
+}
+
+// Define valid player states
+export enum PlayerState {
+ Destroyed = "destroyed", // Player is destroyed by `destroy()` method
+ Error = "error", // An error occurred
+ Loading = "loading", // Player is loading
+ Paused = "paused", // Player is paused
+ Playing = "playing", // Player is playing
+ Stopped = "stopped",  // Player is stopped
+ Frozen = "frozen", // Player is paused due to player being invisible
+}
+
+// Define play modes
+export enum PlayMode {
+  Bounce = "bounce",
+  Normal = "normal",
+}
+
+// Define player events
+export enum PlayerEvent {
+  Complete = "complete",
+  Destroyed = "destroyed",
+  Error = "error",
+  Frame = "frame",
+  Freeze = "freeze",
+  Load = "load",
+  Loop = "loop",
+  Pause = "pause",
+  Play = "play",
+  Ready = "ready",
+  Stop = "stop",
+  }
+
+/**
+ * Parse a resource into a bytes
+ */
+const _parseSrc = async (src: string | object): Promise<Uint8Array> => {
+  let data = src;
+  if (typeof data === "object") {
+    data = JSON.stringify(data);
+  } else if (typeof data === "string") {
+    try {
+      data = JSON.parse(data);
+    } catch (err) {
+      const json = await _fromURL(data as string);
+      data = JSON.stringify(json);
+    }
+  }
+
+  const encoder = new TextEncoder();
+  return encoder.encode(data as string);
+}
+
+const _fromURL = async (url: string): Promise<Record<string, any>> => {
+  if (typeof url !== "string") {
+    throw new Error(`The url value must be a string`);
+  }
+
+  let json;
+
+  try {
+    // Try construct an absolute URL from the src URL
+    const srcUrl: URL = new URL(url);
+
+    // Fetch the JSON file from the URL
+    const result: any = await fetch(srcUrl.toString());
+
+    json = await result.json();
+  } catch (err) {
+    throw new Error(
+      `An error occurred while trying to load the Lottie file from URL`
+    );
+  }
+
+  return json;
+}
```

Comment:
1. Nothing will be executed after throwing an error
2. The creation of `let json;` is unnecessary 
3. Don't need comments when the code explains itself and common among devs
4. If an error has some message - it should be provided to users
5. the function's return type was misleading
6. Moved some validation logic from `_parseSrc` to simplify things
7. Please consider better namings for the function
8. Don't need to include `_` because these are helper functions
9. Ddded better function name
10. Please declare this function before `_parseSrc`, because its been utilized before inilization and you are using the arrow function.
11. please use Type Assertions as little as possible to avoid hidden errors

```suggestion
const getLottieFromUrl = async (url: unknown): Promise<string> => {
  switch (url) {
    case 'object':
      return JSON.stringify(url);
    case 'string':
      break;
    default:
      throw new Error('The url value must be a string');
  }

  try {
    const absoluteUrl = new URL(url);
    const result = await fetch(absoluteUrl.toString());
    const json = await result.json();

    return JSON.stringify(json);
  } catch (err) {
    const errMessage = typeof err === 'string' ? err : '';

    throw new Error(
      `An error occurred while trying to load the Lottie file from URL, ${errMessage}`
    );
  }
};
```

Reply:
> Nothing will be executed after throwing an error

The player error from this case is intercepted in `load()` method.
```ts
try {
  const bytes = await _parseSrc(src);
  this.dispatchEvent(new CustomEvent(PlayerEvent.Ready));

  this._loadBytes(bytes);
} catch (err) {
  this.currentState = PlayerState.Error;
  this.dispatchEvent(new CustomEvent(PlayerEvent.Error));
}
```

Reply:
> Please consider better namings for the function
```ts
const getLottieFromUrl = async (url: unknown): Promise<string>
```

This library and method only parses lottie for now, so I'm not convinced only this method have lottie in the naiming. However original name is not also good enough, will use `parseURL`

Reply:
> Don't need to include _ because these are helper functions
We will have code convention later, for now It'd be better having just uniformity rather than follow specific style.

Reply:
```ts
...
  switch (url) {
    case 'object':
      return JSON.stringify(url);
...
```

There is no sense to check whether `url` is `object` type, literally `url` can't be other types than `string`

Reply:
the only reason I added this case is because you had it in the `_parseSrc` function

Reply:
> We will have code convention later, for now It'd be better having just uniformity rather than follow specific style.

sure it's up to you, but it's famous and common style among JS devs, thus your current styles bring confusion to a reader